packages/fxa-auth-server/scripts/stripe-products-and-plans-to-firestore-documents/stripe-product-and-plans-converter.ts

```diff
@@ -0,0 +1,320 @@
+        const metadataKeyPrefix = `product:${key}`;
+        value = this.getArrayOfStringsFromMetadataKeys(
+          stripeObject.metadata,
+          metadataKeyPrefix
+        );
+        if (value.length > 0) {
+          uiContentConfig[key] = value;
+        }
+        continue;
+      }
+      value =
+        stripeObject.metadata[key] ?? stripeObject.metadata[`product:${key}`];
+      if (value) {
+        uiContentConfig[key] = value;
+      }
+    }
+    return uiContentConfig;
+  }
+
+  // TODO JSDoc
+  urlMetadataToUrlConfig(
+    stripeObject: Stripe.Product | Stripe.Plan
+  ): UrlConfig {
+    const urlConfig: UrlConfig = {};
+    if (!stripeObject.metadata) {
+      return urlConfig;
+    }
+    for (const key of UrlConfigKeys) {
+      const value =
+        stripeObject.metadata[`${key}URL`] ??
+        stripeObject.metadata[`${key}Link`] ??
+        stripeObject.metadata[`product:${key}URL`];
+      if (value) {
+        urlConfig[key] = value;
+      }
+    }
+    return urlConfig;
+  }
+
+  // TODO JSDoc
+  // This method will return a valid ProductConfig, but it will need to be
+  // appended with any locale data for Stripe plans if present.
+  stripeProductToProductConfig(product: Stripe.Product): ProductConfig {
+    // Firestore cannot serialize class instances
+    const productConfig: any = {};
+
+    // BaseConfig
+    productConfig.active = product.active;
+    productConfig.capabilities =
+      this.capabilitiesMetadataToCapabilityConfig(product);
+    // `locales` will be populated once we iterate through the Stripe Plans
+    productConfig.locales = {};
+    productConfig.styles = this.stylesMetadataToStyleConfig(product);
+    productConfig.support = this.supportMetadataToSupportConfig(product);
+    productConfig.uiContent = this.uiContentMetadataToUiContentConfig(product);
+    productConfig.urls = this.urlMetadataToUrlConfig(product);
+
+    // Extended by ProductConfig
+    if (product.id) {
+      productConfig.stripeProductId = product.id;
+    }
+    if (product.metadata.productSet) {
+      productConfig.productSet = product.metadata.productSet;
+    }
+    if (product.metadata.promotionCodes) {
+      productConfig.promotionCodes = commaSeparatedListToArray(
+        product.metadata.promotionCodes
+      );
+    }
+    return productConfig;
+  }
+
+  findLocaleStringFromStripePlan(plan: Stripe.Plan): null | string {
+    // Try to extract a locale from the plan nickname
+    const { nickname } = plan;
+    let locale = nickname
+      ?.split(' ')
+      .filter((w) => this.supportedLanguages.includes(w.toLowerCase()));
+    if (locale && locale.length > 0) {
+      return locale[0];
+    }
+    // Failing that, try to ascertain the language of any localized strings
+    //  - Use Google Translate Client https://cloud.google.com/translate/docs/samples/translate-v3-detect-language#translate_v3_detect_language-nodejs
+    // If the locale is not valid, throw an error
+    // Return the locale
```

Comment:
@chenba : I'd like to file a follow-up ticket for improving the heuristics of this `findLocaleStringFromStripePlan`, since a `productConfig` can be valid with an empty `locales` object, and `locales` is an optional key on the `planConfig` object. Also it seems like a separate, 3 story-point chunk of work. What do you think?

FWIW, I did find a library that infers a language based on a passed string: https://cloud.google.com/translate/docs/samples/translate-v3-detect-language#translate_v3_detect_language-nodejs .

Reply:
> @chenba : I'd like to file a follow-up ticket for improving the heuristics of this

Sounds good.

> FWIW, I did find a library that infers a language based on a passed string ...

Yeah that's the API I had in mind when I jokingly brought the idea up.  I wonder it would return 'it' when we want 'it-ch'; the Swiss locales are the trickier ones.

Reply:
Yeah it might be a little tough for the region parts.

FWIW this one heuristic I already have only extracted the locale from 7 of 41 plans with the current list of locales in `supportedLanguages`.

I filed #12053 ([FXA-4662](https://mozilla-hub.atlassian.net/browse/FXA-4662)).